### ISSUE TITLE
Window not seeing service but Liinux does with `avahi-publish`

### ISSUE THREAD

**Issue by dalesmith3** [keepsimple1/mdns-sd#106]
When publishing a service on Linux with `avahi-publish` like

```
avahi-publish -s Service _http._tcp 8080 path=/
```

The code at https://bitbucket.org/dalepsmith/rust-mdns-discover/src/master/
sees the service on Linux, but not when compiled for Windows.

The windows compiled code *does* see services published via files in `/etc/avahi/services/*.service`

Not sure what's happening, but it's probably related to ordering of received packets?  Or maybe ordering of records in a packet?

**Comment by keepsimple1** [comment]
I tried this with a Raspberry Pi running Linux and then a Windows 11 machine.  With the example query program I was not able to reproduce the issue. (i.e. it worked)
 
```
C:\Users\keeps\work\mdns-sd (main)>  cargo run --example query _http._tcp
    Finished dev [unoptimized + debuginfo] target(s) in 0.05s
     Running `target\debug\examples\query.exe _http._tcp`
At 452.3µs : Received other event: SearchStarted("_http._tcp.local. on addrs [192.168.0.112]")
At 330.4719ms : Received other event: ServiceFound("_http._tcp.local.", "mdns-bug2._http._tcp.local.")
At 330.9278ms: Resolved a new service: mdns-bug2._http._tcp.local. IP: {192.168.0.109} TXT properties: TxtProperties { properties: [TxtProperty {key: "path", val: Some("/")}] }
```

Could you please try to run the example query on Windows and see if it can find the service?  If not, is it possible to add a simple logger to capture the debug log? 

> Not sure what's happening, but it's probably related to ordering of received packets? Or maybe ordering of records in a packet?

There was an issue with the ordering of records in a packet before. Now it should be fine as we always [check for PTR record first](https://github.com/keepsimple1/mdns-sd/blob/3aab8d7f4614d4e53eefecb5b96687f53ed44e33/src/service_daemon.rs#L1325). There is one case when the records are separated into multiple packets and PTR comes in the later packets, we might have issue there.  (btw, I'm looking into refactoring this part of code, so very interested in learning possible issues in its current form).

**Comment by dalepsmith** [comment]
I have not yet run the example query on Windows, but I will today
(need to install Rust dev environment there).

I now suspect the problem I'm is virtual machine / network iface related.

The failing case is window machine on wifi at my home doing mdns
browsing, and ti's failing to see the avahi-publish running in a Linux
vm with the network iface in bridged mode.  It *does* see publishes
from other machine on my local net.  Both via avahi-publish and by
config files in /etc/avahi/services.

When I have the same window machine in the office, it's wired into
Ethernet, and then it *does* see the avahi-publish from the Linux vm.
The vm is vmware workstation player running Debian Stable.

-Dale

**Comment by dalepsmith** [comment]
Ok! Running the example query on windows shows the same thing.
Does not see services published from within the vm,
but does see service published on the local net.

The Linux the vm sees services published on the local net.
Other devices on the local net see the things published by the vm.

So the only failure is with with windows host and the linux guest.
On vmware.  I don't remember this happening with openbox.  But then I always used openbox wired.

Next I need to see if using a wired connection makes any difference.
-Dale

**Comment by dalepsmith** [comment]
[query.txt](https://github.com/keepsimple1/mdns-sd/files/11296510/query.txt)
[foo-query.zip](https://github.com/keepsimple1/mdns-sd/files/11296525/foo-query.zip)
Attached is the example query console output and a zipped wireshark packet capture of the service being removed and then added.

**Comment by keepsimple1** [comment]
> When I have the same window machine in the office, it's wired into Ethernet, and then it *does* see the avahi-publish from the Linux vm.

In this success case, is the network iface also in bridged mode, or not?

**Comment by dalepsmith** [comment]
Yes also in bridged mode.  The vm config is identical.
The difference is the host network connection.

Could it be a difference in "trust"?   Like Corporate lan allows
packets that are blocked/filtered in my home wifi?

**Comment by keepsimple1** [comment]
I tried in a setup of Windows 11 (WiFi home network) and VMware player Debian 11.6 stable, with bridged network iface. But was not able to reproduce the issue.  I plan to add a logger for example code so that it's easier to capture debug logs. 

```
cargo r --example query _http._tcp
    Finished dev [unoptimized + debuginfo] target(s) in 0.05s
     Running `target\debug\examples\query.exe _http._tcp`
At 410.7µs : Received other event: SearchStarted("_http._tcp.local. on addrs [192.168.70.1, 192.168.0.112, 192.168.138.1]")
At 48.877ms : Received other event: ServiceFound("_http._tcp.local.", "Service._http._tcp.local.")
At 49.1775ms: Resolved a new service: Service._http._tcp.local. host: debian-116.local. port: 8080 IP: {192.168.70.128} TXT properties: TxtProperties { properties: [TxtProperty {key: "path", val: Some("/")}] }
```

**Comment by dalepsmith** [comment]
I'll be in the office tomorrow (Wed 26)  to retest.

**Comment by keepsimple1** [comment]
I created a [PR](https://github.com/keepsimple1/mdns-sd/pull/112) to add logging in the examples.  Is it possible to run the query example with this logging support in your home network to reproduce the issue and collect the debug logs? (Or to find out what debug logs are missing?)

**Comment by dalepsmith** [comment]
Yes.  Here is a a log.

[debug-log.txt](https://github.com/keepsimple1/mdns-sd/files/11338290/debug-log.txt)

**Comment by keepsimple1** [comment]
It is odd that there are not PTR or SRV, TXT records received, but only an Address record received: 
```
[2023-04-26T23:58:54Z DEBUG mdns_sd::dns_parser] read_header: id 0, 0 questions 1 answers 0 authorities 0 additionals
[2023-04-26T23:58:54Z DEBUG mdns_sd::dns_parser] read_questions: 0
[2023-04-26T23:58:54Z DEBUG mdns_sd::dns_parser] read_others: 1
[2023-04-26T23:58:54Z DEBUG mdns_sd::dns_parser] DnsAddress { record: DnsRecord { entry: DnsEntry { name: "192-168-1-103.local.", ty: 1, class: 1, unique: true }, ttl: 3600, created: 1682553534969, refresh: 1682556414969 }, address: 192.168.1.103 }
```

Is this address `192.168.1.103` belonging to the published service on Linux?

**Comment by dalepsmith** [comment]
That machine (today) is currently 192.168.1.166. Not sure what is was
when I made the log.
Ok, 103 is an Amazon Fire on the same network.
Cute how the hostname is "192-168-1-103" (note the '-'s)

It *must* be some kind of firewall settings in Windows.

I just tried connected via Ethernet instead of WiFi and things are
working perfectly.
It's still odd that I get browse results from other devices on the
subnet, but not from the vm on the same subnet.

I guess that means we can close this issue.  Thank you for helping me
work through it!

I wonder is there is some setting I can change?
